Databricks adapter 1.11.0 updates (#8136)
fishtown-analytics/docs.getdbt.com: website/docs/reference/global-configs/databricks-changes.md

The following are the current [behavior change flags](/docs/reference/global-configs/behavior-changes.md#behavior-change-flags) that are specific to `dbt-databricks`:

| Flag                          | `dbt-databricks`: Intro | `dbt-databricks`: Maturity | Status |
| ----------------------------- | ----------------------- | -------------------------- | ------ |
| [`use_info_schema_for_columns`](#use-information-schema-for-columns) | 1.9.0                   | N/A                        | **Removed in 1.11.0** |
| [`use_user_folder_for_python`](#use-users-folder-for-python-model-notebooks)  | 1.9.0                   | 1.11.0                     | Default changed to `True` |
| [`use_materialization_v2`](#use-restructured-materializations)      | 1.10.0                  | TBD                        | Active |
| [`use_managed_iceberg`](#use-managed-iceberg)  | 1.11.0  |  1.12.0                                                     | Active |
| [`use_replace_on_for_insert_overwrite`](#use-replace-on-for-insert_overwrite-strategy)   | 1.11.0   | 1.12.0                    | Active, defaults to `True` |

## Use information schema for columns

:::caution Removed in v1.11.0

> The `use_info_schema_for_columns` flag has been **removed** as of dbt-databricks v1.11.0. The adapter now uses `DESCRIBE EXTENDED ... AS JSON` (available in DBR 16.2+) to efficiently retrieve complex type information, eliminating the need for this flag.

Reviewer: adding link to db docs 
```suggestion
The `use_info_schema_for_columns` flag has been **removed** as of dbt-databricks v1.11.0. The adapter now uses [`DESCRIBE EXTENDED ... AS JSON`](https://docs.databricks.com/aws/en/sql/language-manual/sql-ref-syntax-aux-describe-table) (available in DBR 16.2+) to efficiently retrieve complex type information, eliminating the need for this flag.
```

If you're still using this flag in your project configuration, you can safely remove it. The new approach provides better performance and doesn't require the `REPAIR TABLE` operations that were needed with `information_schema`.

:::

> ### Legacy documentation (for versions < 1.11)

Reviewer: ```suggestion
### Legacy documentation
```

> 

Reviewer: ```suggestion
_This applies to dbt-databricks versions v1.11 and older_
```

The `use_info_schema_for_columns` flag was `False` by default in versions 1.9 and 1.10.

Setting this flag to `True` would use `information_schema` rather than `describe extended` to get column metadata for Unity Catalog tables. This setting helped avoid issues where `describe extended` truncates information when the type is a complex struct.

:::tip For complex types

If your complex type comes from processing JSON using `from_json`, you have an alternative: use [`parse_json` to create the column as the `variant` type](https://docs.databricks.com/aws/en/sql/language-manual/functions/parse_json).
The `variant` type might be a reasonable alternative in terms of performance, while avoiding type truncation issues.

:::

## Use user's folder for Python model notebooks

:::info Default changed in v1.11.0

As of dbt-databricks v1.11.0, the `use_user_folder_for_python` flag defaults to **`True`**.

:::

The `use_user_folder_for_python` flag controls where uploaded Python model notebooks are stored in Databricks:

> - **`True` (default in v1.11+)**: Notebooks are written to `/Users/{{current user}}/{{catalog}}/{{schema}}/`

Reviewer: ```suggestion
- **`True` (default in v1.11+)**: Notebooks are written to `/Users/{{current user}}/{{catalog}}/{{schema}}/`.
```

> - **`False` (default in v1.9-v1.10)**: Notebooks are written to `/Shared/dbt_python_models/{{schema}}/`

Reviewer: ```suggestion
- **`False` (default in v1.9-v1.10)**: Notebooks are written to `/Shared/dbt_python_models/{{schema}}/`.
```

> Writing to the `Shared` folder is deprecated by Databricks as it does not align with governance best practices. Using user-specific folders provides better isolation, access control, and aligns with Unity Catalog security models.

Reviewer: ```suggestion
Databricks deprecated writing to the `Shared` folder as it doesn't align with governance best practices. Using user-specific folders provides better isolation, access control, and aligns with Unity Catalog security models.
```

> If you need to preserve the legacy behavior for backward compatibility, you can explicitly set this flag to `False` in your `dbt_project.yml`:

Reviewer: ```suggestion
To preserve the legacy behavior for backward compatibility, you can explicitly set this flag to `False` in your `dbt_project.yml`:
```

```yaml
flags:
  use_user_folder_for_python: false
```

## Use restructured materializations

The `use_materialization_v2` flag is `False` by default and guards significant rewrites of the core materializations in `dbt-databricks` while they are still in an experimental stage.

When set to `True`, `dbt-databricks` uses the updated logic for all model types (views, tables, incremental, seeds). It also enables additional, optional config options for more fine-tuned control:
* `view_update_via_alter` &mdash; When enabled, this config attempts to update the view in place using alter view, instead of using create or replace to replace it.
* `use_safer_relation_operation` &mdash; When enabled (and if `view_update_via_alter` isn't set), this config makes dbt model updates more safe by staging relations and using rename operations to ensure the live version of the table or view is not disrupted by failures.

These configs aren't required to receive the core benefits of this flag &mdash; like better performance and column/constraint functionality &mdash; but they are gated behind the flag because they introduce more significant changes to how materializations behave.

> The default value of this flag remains `False` in v1.11.0.

Reviewer: ```suggestion
In v1.11.0, this flag will stay set to `False` by default. Based on feedback about the new materialization’s lack of atomicity (all-or-nothing updates), we won’t enable it automatically. We’ll explore other ways to achieve the same benefits without losing atomicity.
```

Given feedback about lack of atomicity of the new materialization approach, we will not be flipping this flag to `True`.

> Instead we will be investigating new ways to provide the same benefits while maintaining atomicity.

Reviewer: ```suggestion
Instead, we will be investigating new ways to provide the same benefits while maintaining atomicity.
```
